### ISSUE TITLE
inference time of model_transformed.pb is larger than model_original.pb;although model_transformed.pb is samll than model_original.pb

### ISSUE THREAD

**Issue by distrect9** [Tencent/PocketFlow#95]
**Describe the bug**

**To Reproduce**

**Expected behavior**

**Screenshots**

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Additional context**

**Comment by distrect9** [comment]
the pb file.

**Comment by jiaxiang-wu** [comment]
Similar problem as the issue #83. It is likely that the way we used to transform the channel pruned graph is not efficient on GPU (but it is efficient on ARM chips).

Below is copied from #83:
When creating a transformed graph for the channel pruned model, we decompose each (k x k) convolutional layer into the combination of a 1x1 pointwise convolutional layer (to remove all-zero channels) and a (k x k) convolutional layer (for non-zero channels). The 1x1 pointwise convolutional is not efficient on GPU, so the inference time may be slower. However, according to our experimence, it is more efficient on ARM (maybe also on CPU) chips, since inference time on mobile devices can indeed be shorter.
A possible improvement is to change the way we used above for creating the transformed graph (remove the need for 1x1 pointwise convolutional layer).

**Comment by jiaxiang-wu** [comment]
TODO: optimize the graph editting approach for channel pruning models.

**Comment by distrect9** [comment]
i think it‘s hard to change the graph without adding more computation，because you have to change the previous layer fromat，but the previous layer may have influence on diffent layer，so there seems not to hvae a simple way to change the network architecture;
by the way, the reason for gpu to run slow may be you just add addion computation....  and the zeros parmeter seems can't speed up the caucalation.

**Comment by jiaxiang-wu** [comment]
@distrect9 
After channel pruning, we have two options in modifying the graph:
1. tf.gather + standard conv: use tf.gather to filter out all-zero channels, and then apply the standard convolutional operation. However, tf.gather is not fully supported by TensorFlow Lite, and cannot be used for TF-Lite based deployment.
2. 1x1 conv + standard conv: use 1x1 conv to obtain feature maps with non-zero channels only, and then apply the standard convolutional operation. This will introduce extra computation due to 1x1 conv and slow the compressed model down.

I think a possible approach is to adopt scheme 1 for GPU-based computation (also, the data format may need to change into NCHW) and adopt scheme 2 for TensorFlow Lite deployment.

**Comment by distrect9** [comment]
yes, for scheme 1, it looks like that we have to change the net code and the channel pruned code.....   will you do this?

**Comment by jiaxiang-wu** [comment]
@distrect9 Looking forward to hearing good news from you.

**Comment by jiaxiang-wu** [comment]
@distrect9 What is the data format? NHWC or NCHW?

**Comment by distrect9** [comment]
NHWC

**Comment by distrect9** [comment]
I just change that line

**Comment by distrect9** [comment]
maybe i just test on cifiar10, the pic size is too small, i'll try on imagenet.

**Comment by jiaxiang-wu** [comment]
@distrect9 How do you change the data format to NHWC? Can you post your model conversion code?

**Comment by distrect9** [comment]
the source type is NHWC, i did't change the training format.
the tools code is in https://github.com/distrect9/for_pocket_flow

**Comment by distrect9** [comment]
i just change one line and delete the tf-lite format

**Comment by jiaxiang-wu** [comment]
Sorry for the typo. You may need to convert the data format into NCHW during model conversion, since tf.gather() may be faster on NCHW than NHWC.

**Comment by distrect9** [comment]
ok，i will try it.

**Comment by distrect9** [comment]
i test tf.gather()  on different axis. it did consum less time when axis = 1 than when axis=3.  when axis is 1,for 1000 time calculate,it cost about 1.68s,and when axis is 3, it cost 0.99s.

**Comment by distrect9** [comment]
i'll test on model

**Comment by distrect9** [comment]
btw, you go to work so early......

**Comment by jiaxiang-wu** [comment]
@distrect9 Sorry, could you please clarify your results on using tf.gather() on different axis? You said "it did consum less time when axis = 1 than when axis=3", but later you mentioned that with 1000 runs, for axis=1 the time is 1.68s and for axis=3 the time is 0.99s (shorter?).

**Comment by distrect9** [comment]
my fault,it should be:
i test tf.gather() on different axis. it did consum less time when axis = 1 than when axis=3. when axis is 1,for 1000 time calculate,it cost about 0.99s,and when axis is 3, it cost 1.68s.

**Comment by jiaxiang-wu** [comment]
That's a good news. Thus, it is promising that using tf.gather and NCHW data format should improve the inference speed on GPUs than the current implementation (1x1 conv + NHWC).

**Comment by distrect9** [comment]
i can't find the args for discr_channel_pruning  to change the pruned channel index, do i need to change the code my self?

**Comment by jiaxiang-wu** [comment]
What do you mean by "change the pruned channel index"?

**Comment by distrect9** [comment]
yes

**Comment by distrect9** [comment]
yes

**Comment by distrect9** [comment]
yes

**Comment by jiaxiang-wu** [comment]
@distrect9 What do you mean by "change the pruned channel index"?

**Comment by jiaxiang-wu** [comment]
No need to modify the DCP algorithm's code, You can change the data format using this script: `tools/conversion/convert_data_format.py`.

**Comment by distrect9** [comment]
use tools/conversion/convert_data_format.py  to process pruned pb file?

**Comment by jiaxiang-wu** [comment]
No, this script is used to convert checkpoint files in NHWC to checkpoint files in NCHW, or vice versa.

**Comment by distrect9** [comment]
ok, i know

**Comment by distrect9** [comment]
sad news：the pruned pb is still slower than original pb.
but good news: 

  | resnet  NHWC | resnet NCHW| 
origin | 19.533841609954834 | 22.28275775909424
after pruned | 25.992661237716675 | 24.99310612678528

it did speed up for pruned compared with NHWC.

btw,tonight i'll get the results on ilsvrc12 ,and i think on ilsvrc12 ,the pruned pb will be faster than original pb.....

**Comment by jiaxiang-wu** [comment]
@distrect9 We have released some new results on ILSVRC-12 (tf.gather + conv, NCHW) in this pull request. Basically, this is helpful on ResNet-like models. MobileNet does not suppport NCHW, so we did not test on it.

**Comment by distrect9** [comment]
ok，thanks
